File: app/src/main/java/com/example/waveplayer/MediaPlayerWUri.java
package com.example.waveplayer;

import android.media.MediaPlayer;
import android.os.Build;
import android.util.Log;

import static android.media.MediaPlayer.SEEK_CLOSEST;

public class MediaPlayerWUri {

    private static final String TAG = "MediaPlayerWURI";

    private final Object lock = new Object();

    private MediaPlayer mediaPlayer;

    final AudioUri audioURI;

    private final ServiceMain serviceMain;

    private volatile boolean isPrepared = true;

    public boolean isPrepared() {
        synchronized (lock) {
            return isPrepared;
        }
    }

    private volatile boolean shouldPlay = false;

    MediaPlayerWUri(final ServiceMain serviceMain, MediaPlayer mediaPlayer, AudioUri audioURI){
        this.serviceMain = serviceMain;
        this.mediaPlayer = mediaPlayer;
        this.audioURI = audioURI;
        mediaPlayer.setOnPreparedListener(null);
        mediaPlayer.setOnErrorListener(null);
        MOnPreparedListener mOnPreparedListener = new MOnPreparedListener(this);
        mediaPlayer.setOnPreparedListener(mOnPreparedListener);
        MediaPlayer.OnErrorListener onErrorListener = new MediaPlayer.OnErrorListener() {
            @Override
            public boolean onError(MediaPlayer mediaPlayer, int i, int i1) {
                synchronized (lock) {
                    serviceMain.releaseMediaPlayers();
                    serviceMain.addToQueueAndPlay(serviceMain.getCurrentSong().getUri());
                    return false;
                }
            }
        };
        mediaPlayer.setOnErrorListener(onErrorListener);
    }

    public void shouldPlay(boolean shouldPlay){
        synchronized (lock) {
            if (shouldPlay && isPrepared) {
                mediaPlayer.start();
            } else {
                this.shouldPlay = shouldPlay;
            }
        }
    }

    public void release(){
        synchronized (lock) {
            isPrepared = false;
            shouldPlay = false;
            mediaPlayer.reset();
            mediaPlayer.release();
        }
    }

    public void prepareAsync(){
        Log.v(TAG, "Waiting for prepareAsync");
        synchronized (lock) {
            isPrepared = false;
            mediaPlayer.prepareAsync();
        }
    }

    public int getCurrentPosition(){
        synchronized (lock) {
            if (isPrepared && serviceMain.songInProgress()) {
                return mediaPlayer.getCurrentPosition();
            } else {
                return -1;
            }
        }
    }

    public boolean isPlaying(){
        synchronized (lock) {
            if (isPrepared) {
                return mediaPlayer.isPlaying();
            } else {
                return false;
            }
        }
    }

    public void stop(){
        Log.v(TAG, "MediaPlayer stopped");
        synchronized (lock) {
            isPrepared = false;
            shouldPlay = false;
            mediaPlayer.stop();
        }
    }

    public void pause(){
        synchronized (lock) {
            if (isPrepared) {
                mediaPlayer.pause();
            }
            shouldPlay = false;
        }
    }

    public void seekTo(int millis){
        synchronized (lock) {
            if (isPrepared) {
                String[] s = audioURI.displayName.split("\\.");
                if(s.length > 0) {
                    if(s[s.length - 1].toLowerCase().equals("mkv")){
                        mediaPlayer.reset();
                        mediaPlayer.release();
                        mediaPlayer = null;
                        mediaPlayer = MediaPlayer.create(
                                serviceMain.getApplicationContext(), audioURI.getUri());
                        mediaPlayer.setOnPreparedListener(null);
                        mediaPlayer.setOnErrorListener(null);
                        MOnPreparedListener mOnPreparedListener = new MOnPreparedListener(this);
                        mediaPlayer.setOnPreparedListener(mOnPreparedListener);
                        MediaPlayer.OnErrorListener onErrorListener = new MediaPlayer.OnErrorListener() {
                            @Override
                            public boolean onError(MediaPlayer mediaPlayer, int i, int i1) {
                                synchronized (lock) {
                                    serviceMain.releaseMediaPlayers();
                                    serviceMain.addToQueueAndPlay(serviceMain.getCurrentSong().getUri());
                                    return false;
                                }
                            }
                        };
                        mediaPlayer.setOnErrorListener(onErrorListener);
                        setOnCompletionListener(serviceMain.onCompletionListener);
                        this.shouldPlay = true;
                        return;
                    }
                }
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    mediaPlayer.seekTo(millis, SEEK_CLOSEST);
                } else {
                    mediaPlayer.seekTo(millis);
                }
            }
        }
    }

    public void setOnCompletionListener(MediaPlayer.OnCompletionListener onCompletionListener) {
        mediaPlayer.setOnCompletionListener(onCompletionListener);
    }

    class MOnPreparedListener implements MediaPlayer.OnPreparedListener{

        private final MediaPlayerWUri mediaPlayerWURI;

        public MOnPreparedListener(MediaPlayerWUri mediaPlayerWURI){
            this.mediaPlayerWURI = mediaPlayerWURI;
        }

        @Override
        public void onPrepared(MediaPlayer mediaPlayer) {
            synchronized (mediaPlayerWURI) {
                Log.v(TAG, "MediaPlayer prepared");
                mediaPlayerWURI.isPrepared = true;
                if (shouldPlay) {
                    mediaPlayer.start();
                    shouldPlay = false;
                }
            }
        }
    }

}

File: app/src/main/java/io/fourth_finger/pinky_player/random_playlist/MediaPlayerWUri.kt
package io.fourth_finger.pinky_player.random_playlist

import android.content.Context
import android.media.MediaPlayer
import android.media.MediaPlayer.OnPreparedListener
import android.os.Build
import io.fourth_finger.playlist_data_source.AudioUri
import java.util.*

class MediaPlayerWUri(
    private val mediaPlayerManager: MediaPlayerManager,
    private var mediaPlayer: MediaPlayer,
    val id: Long
) {

    @Volatile
    private var isPrepared: Boolean = false
    fun isPrepared(): Boolean {
        synchronized(lock) { return isPrepared }
    }

    @Volatile
    private var shouldPlay: Boolean = false

    fun shouldPlay(shouldPlay: Boolean) {
        synchronized(lock) {
            if (shouldPlay && isPrepared) {
                mediaPlayer.start()
            } else {
                this.shouldPlay = shouldPlay
            }
        }
    }

    fun release() {
        synchronized(lock) {
            mediaPlayer.setOnCompletionListener(null)
            isPrepared = false
            shouldPlay = false
            mediaPlayer.reset()
            mediaPlayer.release()
        }
    }

    fun prepareAsync() {
        synchronized(lock) {
            isPrepared = false
            mediaPlayer.prepareAsync()
        }
    }

    fun getCurrentPosition(): Int {
        synchronized(lock) {
            return if (isPrepared) {
                mediaPlayer.currentPosition
            } else {
                -1
            }
        }
    }

    fun isPlaying(): Boolean {
        synchronized(lock) {
            return if (isPrepared) {
                mediaPlayer.isPlaying
            } else {
                false
            }
        }
    }

    fun stop(
        context: Context,
        audioUri: AudioUri
    ) {
        synchronized(lock) {
            isPrepared = false
            shouldPlay = false
            mediaPlayer.stop()
            resetIfMKV(audioUri, context)
        }
    }

    fun pause() {
        synchronized(lock) {
            if (isPrepared) {
                mediaPlayer.pause()
            }
            shouldPlay = false
        }
    }

    fun seekTo(
        context: Context,
        audioUri: AudioUri,
        millis: Int
    ) {
        synchronized(lock) {
            if (isPrepared) {
                if (resetIfMKV(audioUri, context)) {
                    shouldPlay = true
                    return
                }
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    mediaPlayer.seekTo(millis.toLong(), MediaPlayer.SEEK_CLOSEST)
                } else {
                    mediaPlayer.seekTo(millis)
                }
            }
        }
    }

    private fun resetIfMKV(
        audioUri: AudioUri,
        context: Context
    ): Boolean {
        val s: Array<String> = audioUri.displayName.split("\\.").toTypedArray()
        if (s.isNotEmpty()) {
            if ((s[s.size - 1].lowercase(Locale.ROOT) == "mkv")) {
                mediaPlayer.reset()
                mediaPlayer.release()
                mediaPlayer = MediaPlayer.create(context, audioUri.getUri())
                mediaPlayer.setOnPreparedListener(null)
                mediaPlayer.setOnErrorListener(null)
                mediaPlayer.setOnPreparedListener(mOnPreparedListener)
                mediaPlayer.setOnErrorListener(mediaPlayerManager.onErrorListener)
                mediaPlayer.setOnCompletionListener(mediaPlayerManager.onCompletionListener)
                return true
            }
        }
        return false
    }

    private val mOnPreparedListener = OnPreparedListener { mediaPlayer ->
        synchronized(this) {
            isPrepared = true
            if (shouldPlay) {
                mediaPlayer.setOnCompletionListener(mediaPlayerManager.onCompletionListener)
                mediaPlayer.start()
                shouldPlay = false
            }
        }
    }

    companion object {
        val lock: Any = Any()
    }

    init {
        mediaPlayer.setOnPreparedListener(null)
        mediaPlayer.setOnErrorListener(null)
        mediaPlayer.setOnPreparedListener(mOnPreparedListener)
        mediaPlayer.setOnErrorListener(mediaPlayerManager.onErrorListener)
    }
}